File: src/types/desktop.ts
/**
 * Desktop Cleanup types (development-plan.md sections 38, 66, 67, 81).
 *
 * Every interface here mirrors a struct in
 * `src-tauri/src/services/desktop.rs` field for field, snake_case included, so
 * a value crossing `invoke` needs no reshaping in either direction — the same
 * contract `src/types/downloads.ts` keeps.
 *
 * The kind and age ids are the wire values: `DesktopKind` and `AgeBucket` in
 * Rust serialise to exactly these strings, so a bucket is called the same
 * thing in the scan, across `invoke`, and in the summary panel.
 */

/**
 * What one thing on the desktop is.
 *
 * The five buckets Downloads has, plus the two a desktop has and a downloads
 * folder does not: a shortcut is a `.lnk` or `.url` reported at its own size,
 * and a folder is counted rather than opened.
 */
export const DESKTOP_KIND_IDS = [
  "shortcut",
  "folder",
  "image",
  "document",
  "archive",
  "installer",
  "other",
] as const;

export type DesktopKindId = (typeof DESKTOP_KIND_IDS)[number];

/**
 * The four rolling windows an item's last-modified time falls in — the same
 * ones the screenshot organizer uses, measured against the same clock.
 *
 * Each item lands in the *narrowest* window that contains it, so the four
 * counts add up to the item count rather than nesting the way section 42's
 * mockup does.
 */
export const AGE_BUCKET_IDS = ["today", "this_week", "this_month", "older"] as const;

export type AgeBucketId = (typeof AGE_BUCKET_IDS)[number];

/**
 * Which of the two desktops something is on.
 *
 * `public` is the All Users desktop Windows composites into the same screen.
 * It is read and labelled and never written to: that needs elevation this app
 * does not ask for.
 */
export type DesktopRootKind = "user" | "public";

/** One thing at the top level of a desktop folder. */
export interface DesktopItem {
  /**
   * Absolute path, and the handle an action sends back.
   *
   * Rust re-checks it before touching anything — see `resolve_item` — so this
   * is a reference the UI passes around rather than a permission it holds.
   */
  path: string;
  name: string;
  /** Lowercased, without the dot. Null for a folder or a file that has none. */
  extension: string | null;
  kind: DesktopKindId;
  /**
   * The item's own size. For a shortcut that is the size of the `.lnk` itself,
   * never of what it points at; for a folder it is 0, because measuring a
   * folder would mean walking it and this scan does not walk folders.
   */
  size_bytes: number;
  /**
   * How many things are directly inside, for a folder. Null for everything
   * that is not one, and for a folder that could not be read.
   */
  item_count: number | null;
  /** Seconds since the Unix epoch, or null on a filesystem that would not say. */
  modified_at: number | null;
  age: AgeBucketId;
  /** The label of the desktop it was found on: "Desktop", "Public Desktop". */
  source: string;
  /**
   * Whether Move and Delete may be offered for it at all.
   *
   * False for every folder and for everything on the public desktop. The row
   * draws without a checkbox when this is false; Rust refuses those paths
   * again regardless, because a disabled checkbox is a decision made in a
   * component and the one that counts is made in the service.
   */
  actionable: boolean;
}

export interface KindCount {
  kind: DesktopKindId;
  /** The summary's label, sent by Rust so it is written once. */
  label: string;
  count: number;
  /** Always 0 for folders, which are never measured. */
  size_bytes: number;
}

export interface AgeCount {
  age: AgeBucketId;
  label: string;
  count: number;
  size_bytes: number;
}

/** One desktop folder's contribution, reported whether or not it had anything. */
export interface DesktopSource {
  label: string;
  path: string;
  kind: DesktopRootKind;
  /** False when the folder is simply not there — normal off Windows. */
  exists: boolean;
  /** Whether anything found here can be moved or deleted. */
  actionable: boolean;
  item_count: number;
  /** Why the folder could not be read. Null when it read fine or is absent. */
  error: string | null;
}

export interface DesktopScan {
  /** Every folder that was looked in, in the order they were read. */
  sources: DesktopSource[];
  item_count: number;
  /** Ordinary files: not shortcuts, not folders. */
  file_count: number;
  shortcut_count: number;
  folder_count: number;
  /** Bytes across everything that has a size. Folders contribute nothing. */
  total_bytes: number;
  /** All seven kinds in summary order, empty ones included. */
  kinds: KindCount[];
  /** All four age windows, newest first, empty ones included. */
  ages: AgeCount[];
  /** The largest item that has a size. Null for a desktop of only folders. */
  largest: DesktopItem | null;
  /** Everything found, largest first. */
  items: DesktopItem[];
  /**
   * Entries that could not be read, plus symlinks, junctions and other reparse
   * points. Reported, never listed, never fatal.
   */
  unreadable_count: number;
}

/** What happened to one item in a move or a delete. */
export interface ItemActionOutcome {
  path: string;
  name: string;
  ok: boolean;
  /** A sentence worth showing. Null when `ok`. */
  error: string | null;
  /** Where it ended up, for a move that worked. */
  moved_to: string | null;
}

/**
 * The result of one confirmed action.
 *
 * Every requested item gets an entry whether or not it worked: a locked file
 * half way down a selection should be named, not swallowed.
 */
export interface ActionReport {
  succeeded: number;
  failed: number;
  /** Bytes actually moved or freed — the successes only. */
  bytes: number;
  outcomes: ItemActionOutcome[];
}

/**
 * The two things the review flow can do to an item.
 *
 * Kept as a type rather than a boolean because the confirmation dialog says
 * something different for each, and the difference — one is recoverable, one
 * is not — is the whole reason section 67 exists.
 */
export type DesktopAction = "move" | "delete";

File: src-tauri/src/commands/desktop.rs
//! Commands for Desktop Cleanup (development-plan.md sections 38, 66, 67, 81).
//!
//! Thin, per section 86: [`services::desktop`](crate::services::desktop) holds
//! every rule about what may be touched, and what lives here is the one thing
//! it cannot know — *where the desktops are* — because that is a different
//! path on every OS and only Tauri's path resolver and the environment can
//! answer it.
//!
//! Resolving the folders is deliberately the *only* way they are chosen. No
//! command here takes a desktop folder from the frontend, so there is no
//! argument a bug or a stray `invoke` could set to make the cleanup tools act
//! somewhere else on the disk. `destination` on [`move_desktop_items`] is the
//! single path that comes in from outside, it can only ever be a move
//! *target*, and the service checks it exists, is a directory, is not a
//! desktop itself and is not on the public one before anything is written
//! into it.
//!
//! Section 67's order — scan, show, select, confirm, act — is kept by what is
//! and is not exposed here: [`scan_desktop`] reads and changes nothing, and
//! the two that do change something take an explicit list of items. There is
//! no command that clears "everything older than a month" or "all the
//! shortcuts", because there is no point in the product where the app, rather
//! than the user, gets to decide that — and so no quest, schedule or tray item
//! has anything to call.
//!
//! The two that change something also leave a row in `cleanup_actions` once
//! at least one item was handled, for the cleanup quest and the Organized
//! achievement to count. The row is written after the items have been
//! handled, never before, and holds no path.

use std::path::PathBuf;

use tauri::{AppHandle, Manager, State};

use crate::db::DbConnection;
use crate::services::cleanup_actions::{self, CleanupAction, CleanupUtility};
use crate::services::desktop::{self, ActionReport, DesktopRoot, DesktopRootKind, DesktopScan};
use crate::services::screenshots::{now_epoch, LocalClock};

/// Reads both desktops and describes what is on them.
///
/// The database is touched once, for the local UTC offset the age buckets are
/// measured against, and the lock is released before the walk starts — a
/// desktop full of files takes a moment, and holding the shared connection
/// across it would stall every other query in the app for no reason.
///
/// Step one of section 67, and the only step that happens without the user
/// having selected anything. It changes nothing.
#[tauri::command]
pub fn scan_desktop(app: AppHandle, db: State<DbConnection>) -> Result<DesktopScan, String> {
    let clock = {
        let conn = db.lock().map_err(|error| error.to_string())?;
        LocalClock::from_db(&conn).map_err(|error| error.to_string())?
    };

    let roots = desktop_roots(&app);
    if roots.is_empty() {
        return Err("Could not find your Desktop folder on this computer.".to_string());
    }

    Ok(desktop::scan(&roots, clock, now_epoch()))
}

/// Moves the items the user selected into a folder they picked.
///
/// The recoverable action, and the one the review flow offers first: nothing
/// is lost, so something that turns out to have been wanted is still there.
#[tauri::command]
pub fn move_desktop_items(
    app: AppHandle,
    db: State<DbConnection>,
    paths: Vec<String>,
    destination: String,
) -> Result<ActionReport, String> {
    let report = desktop::move_items(&desktop_roots(&app), &paths, &destination)
        .map_err(|error| error.to_string())?;
    cleanup_actions::note(
        &db,
        CleanupUtility::Desktop,
        CleanupAction::Move,
        report.succeeded,
    );
    Ok(report)
}

/// Deletes the items the user selected, permanently.
///
/// The last step of section 67's chain. It is only reached from a confirmation
/// dialog that names the count and says there is no undo, and it never runs on
/// a list the user did not tick: an empty `paths` is refused by the service
/// rather than treated as "all of them". Folders and anything on the public
/// desktop are refused there too, whatever the frontend sends.
#[tauri::command]
pub fn delete_desktop_items(
    app: AppHandle,
    db: State<DbConnection>,
    paths: Vec<String>,
) -> Result<ActionReport, String> {
    let report =
        desktop::delete_items(&desktop_roots(&app), &paths).map_err(|error| error.to_string())?;
    cleanup_actions::note(
        &db,
        CleanupUtility::Desktop,
        CleanupAction::Delete,
        report.succeeded,
    );
    Ok(report)
}

/// Opens one item with whatever the OS opens it with.
///
/// Not part of section 67's chain — it changes nothing — but it is most of how
/// somebody decides. "Is this the shortcut I still use?" is answered by
/// looking, and having to leave the app to look is how a review gets abandoned
/// half-finished. Guarded by the same [`desktop::resolve_item`] check the
/// destructive commands use, so this cannot become a way to launch an
/// arbitrary path.
#[tauri::command]
pub fn open_desktop_item(app: AppHandle, path: String) -> Result<(), String> {
    let resolved =
        desktop::resolve_item(&desktop_roots(&app), &path).map_err(|error| error.to_string())?;

    tauri_plugin_opener::open_path(resolved.to_string_lossy().into_owned(), None::<&str>)
        .map_err(|error| format!("Could not open {path}: {error}"))
}

/// Shows one item in the OS file manager.
///
/// The other half of "let me look first", and the safer of the two for a
/// shortcut: revealing a `.lnk` shows the shortcut, where opening one runs
/// whatever it points at.
#[tauri::command]
pub fn reveal_desktop_item(app: AppHandle, path: String) -> Result<(), String> {
    let resolved =
        desktop::resolve_item(&desktop_roots(&app), &path).map_err(|error| error.to_string())?;

    tauri_plugin_opener::reveal_item_in_dir(&resolved)
        .map_err(|error| format!("Could not show {path} in the file manager: {error}"))
}

/// The desktops to read, in the order they are reported.
///
/// Two of them, because Windows composites both into the one surface the user
/// is looking at:
///
/// * the signed-in user's own Desktop, from Tauri's path resolver, which is
///   the only folder anything is ever done to;
/// * the Public (All Users) Desktop, from `%PUBLIC%` with the standard
///   location as a fallback — read and labelled so the scan agrees with what
///   is on screen, and refused by the service for every action, because
///   writing there needs elevation this app does not ask for.
///
/// A folder that is not there is simply left off the list and the scan reports
/// it as missing, which is the normal case for the public desktop off Windows.
/// Duplicates are dropped: a redirected or roamed Desktop can make both
/// entries the same folder, and a second copy of every item under a second
/// label would be a confusing way to say "there is only one".
fn desktop_roots(app: &AppHandle) -> Vec<DesktopRoot> {
    let mut roots: Vec<DesktopRoot> = Vec::new();

    if let Ok(desktop) = app.path().desktop_dir() {
        push_root(&mut roots, "Desktop", desktop, DesktopRootKind::User);
    }

    if let Some(public) = public_desktop_dir() {
        push_root(
            &mut roots,
            "Public Desktop",
            public,
            DesktopRootKind::Public,
        );
    }

    roots
}

/// Where Windows keeps the desktop every account shares.
///
/// `%PUBLIC%` is what the OS itself uses and is what a redirected profile
/// updates, so it is read first; `C:\Users\Public` is the fallback for the
/// rare shell that does not set it. There is no equivalent anywhere else, so
/// off Windows this is simply `None` and the scan reports one desktop.
#[cfg(windows)]
fn public_desktop_dir() -> Option<PathBuf> {
    let public = std::env::var_os("PUBLIC")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from(r"C:\Users\Public"));

    Some(public.join("Desktop"))
}

#[cfg(not(windows))]
fn public_desktop_dir() -> Option<PathBuf> {
    None
}

/// Adds a root unless the same folder is already on the list under another
/// name. Compared on the canonical path so a junction, a symlink or a
/// redirected profile folder resolves to whatever it actually points at.
fn push_root(roots: &mut Vec<DesktopRoot>, label: &str, path: PathBuf, kind: DesktopRootKind) {
    let key = canonical_key(&path);
    if roots.iter().any(|root| canonical_key(&root.path) == key) {
        return;
    }
    roots.push(DesktopRoot::new(label, path, kind));
}

/// The canonical form of `path`, or the path itself when it does not exist —
/// which still compares correctly against another spelling of the same
/// missing folder.
fn canonical_key(path: &PathBuf) -> PathBuf {
    std::fs::canonicalize(path).unwrap_or_else(|_| path.clone())
}
